crates/red_knot_workspace/src/watch/workspace_watcher.rs

```diff
@@ -32,53 +41,76 @@ impl WorkspaceWatcher {
     }
 
     pub fn update(&mut self, db: &RootDatabase) {
-        let new_watch_paths = db.workspace().paths_to_watch(db);
+        let search_paths: Vec<_> = system_module_search_paths(db.upcast()).collect();
+        let workspace_path = db.workspace().root(db).to_path_buf();
 
-        let mut added_folders = new_watch_paths.difference(&self.watched_paths).peekable();
-        let mut removed_folders = self.watched_paths.difference(&new_watch_paths).peekable();
+        let new_cache_key = Self::compute_cache_key(&workspace_path, &search_paths);
 
-        if added_folders.peek().is_none() && removed_folders.peek().is_none() {
+        if self.cache_key == Some(new_cache_key) {
             return;
         }
 
-        for added_folder in added_folders {
-            // Log a warning. It's not worth aborting if registering a single folder fails because
-            // Ruff otherwise stills works as expected.
-            if let Err(error) = self.watcher.watch(added_folder) {
-                // TODO: Log a user-facing warning.
-                tracing::warn!("Failed to setup watcher for path '{added_folder}': {error}. You have to restart Ruff after making changes to files under this path or you might see stale results.");
-                self.errored_paths.push(added_folder.clone());
+        // Unregister all watch paths because ordering is important for linux because
+        // it only emits an event for the last added watcher if a subtree is covered by multiple watchers.
+        // A path can be covered by multiple watchers if symlinks are involved.
```

Comment:
I'm not sure I understand how this can happen if a) we first canonicalize all paths and then b) we de-deduplicate nested paths.

Reply:
I added an example.